### PR TITLE
Included SQL License Type and Resource Tagging Feature

### PR DESCRIPTION
**Pull request by singhabh27**
## Checklist

- [ ] I have read the [_Submitting Changes_](../blob/master/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](../blob/master/CONTRIBUTING.md)
- [ ] The title of the PR is clear and informative
- [ ] The appropriate `ChangeLog.md` file(s) has been updated:
    - For any service, the `ChangeLog.md` file can be found at `src/{{SERVICE}}/{{SERVICE}}/ChangeLog.md`
    - A snippet outlining the change(s) made in the PR should be written under the `## Upcoming Release` header -- no new version header should be added
- [ ] The PR does not introduce [breaking changes](../blob/master/documentation/breaking-changes/breaking-changes-definition.md)
- [ ] If applicable, the changes made in the PR have proper test coverage
- [ ] For public API changes to cmdlets:
    - [ ] a cmdlet design review was approved for the changes in [this repository](https://github.com/Azure/azure-powershell-cmdlet-review-pr) (_Microsoft internal only_)
        - {Please put the link here}
    - [ ] the markdown help files have been regenerated using the commands listed [here](../blob/master/documentation/development-docs/help-generation.md#updating-all-markdown-files-in-a-module)

### CODE REVIEW

File: src/Migrate/custom/New-AzMigrateServerReplication.ps1 [hunk 1]
```diff
@@ -119,6 +127,36 @@ function New-AzMigrateServerReplication {
...
+        [Parameter()]
+        [Microsoft.Azure.PowerShell.Cmdlets.Migrate.Category('Path')]
+        [Microsoft.Azure.PowerShell.Cmdlets.Migrate.Models.Api20210210.IVMwareCbtEnableMigrationInputTargetDiskTags]
+        # Specifies the tag to be used for Disk creation.
+        ${DiskTag},
+
+        [Parameter()]
+        [Microsoft.Azure.PowerShell.Cmdlets.Migrate.Category('Path')]
+        [Microsoft.Azure.PowerShell.Cmdlets.Migrate.Models.Api20210210.IVMwareCbtEnableMigrationInputSeedDiskTags]
+        # Specifies the tag to be used for Seed Disk creation.
```
Comment (prsadhu-ms-idc) [inline src/Migrate/custom/New-AzMigrateServerReplication.ps1:151]: seed disk

File: src/Migrate/custom/New-AzMigrateServerReplication.ps1 [hunk 2]
```diff
@@ -119,6 +127,36 @@ function New-AzMigrateServerReplication {
...
+        [Parameter()]
+        [Microsoft.Azure.PowerShell.Cmdlets.Migrate.Category('Path')]
+        [Microsoft.Azure.PowerShell.Cmdlets.Migrate.Models.Api20210210.IVMwareCbtEnableMigrationInputTargetNicTags]
+        # Specifies the tag to be used for Nic creation.
+        ${NicTag},
+
+        [Parameter()]
+        [Microsoft.Azure.PowerShell.Cmdlets.Migrate.Category('Path')]
+        [Microsoft.Azure.PowerShell.Cmdlets.Migrate.Models.Api20210210.IVMwareCbtEnableMigrationInputTargetDiskTags]
+        # Specifies the tag to be used for Disk creation.
```
Comment (prsadhu-ms-idc) [inline src/Migrate/custom/New-AzMigrateServerReplication.ps1:145]: disk

File: src/Migrate/custom/New-AzMigrateServerReplication.ps1 [hunk 3]
```diff
@@ -119,6 +127,36 @@ function New-AzMigrateServerReplication {
...
+        [Parameter()]
+        [Microsoft.Azure.PowerShell.Cmdlets.Migrate.Category('Path')]
+        [Microsoft.Azure.PowerShell.Cmdlets.Migrate.Models.Api20210210.IVMwareCbtEnableMigrationInputTargetVmtags]
+        # Specifies the tag to be used for VM creation.
+        ${VMTag},
+
+        [Parameter()]
+        [Microsoft.Azure.PowerShell.Cmdlets.Migrate.Category('Path')]
+        [Microsoft.Azure.PowerShell.Cmdlets.Migrate.Models.Api20210210.IVMwareCbtEnableMigrationInputTargetNicTags]
+        # Specifies the tag to be used for Nic creation.
```
Comment (prsadhu-ms-idc) [inline src/Migrate/custom/New-AzMigrateServerReplication.ps1:139]: NIC

File: src/Migrate/custom/New-AzMigrateServerReplication.ps1 [hunk 4]
```diff
@@ -54,6 +54,14 @@ function New-AzMigrateServerReplication {
         # Specifies if Azure Hybrid benefit is applicable for the source server to be migrated.
         ${LicenseType},
 
+        [Parameter()]
+        [ValidateSet("NoLicenseType" , "PAYG" , "AHUB")]
+        [ArgumentCompleter( { "NoLicenseType" , "PAYG" , "AHUB" })]
+        [Microsoft.Azure.PowerShell.Cmdlets.Migrate.Category('Path')]
+        [System.String]
+        # Specifies if Azure Hybrid benefit is applicable for the source server to migrated.
```
Comment (prsadhu-ms-idc) [inline src/Migrate/custom/New-AzMigrateServerReplication.ps1:62]: Review this with Rahul Gupta PM.

File: src/Migrate/custom/New-AzMigrateServerReplication.ps1 [hunk 5]
```diff
@@ -464,6 +514,251 @@ public static int hashForArtifact(String artifact)
         if ($HasTargetAVZone) {
             $ProviderSpecificDetails.TargetAvailabilityZone = $TargetAvailabilityZone
         }
+
+        if ($HasSqlServerLicenseType)
+        {
+            $validLicenseSpellings = @{
```
Comment (prsadhu-ms-idc) [inline src/Migrate/custom/New-AzMigrateServerReplication.ps1:520]: SqlLicense

File: src/Migrate/custom/New-AzMigrateServerReplication.ps1 [hunk 6]
```diff
@@ -464,6 +514,251 @@ public static int hashForArtifact(String artifact)
...
+            $SqlServerLicenseType = $validLicenseSpellings[$SqlServerLicenseType]
+            $ProviderSpecificDetails.SqlServerLicenseType = $SqlServerLicenseType
+        }
+
+        if ($HasTag) {
+            $IllegalCharKey = New-Object Collections.Generic.List[String]
+            $ExceededLengthKey = New-Object Collections.Generic.List[String]
+            $ExceededLengthValue = New-Object Collections.Generic.List[String]
+
+            if ($Tag.Count -gt 50)
```
Comment (prsadhu-ms-idc) [inline src/Migrate/custom/New-AzMigrateServerReplication.ps1:534]: Checking tag.count, but prining Vmtag.count.. please check.

File: src/Migrate/custom/New-AzMigrateServerReplication.ps1 [hunk 7]
```diff
@@ -464,6 +514,251 @@ public static int hashForArtifact(String artifact)
...
+            {
+                throw "InvalidTags : Too many tags specified. Requested tag count - '$($VMTag.Count)'. Maximum number of tags allowed - '50'."
+            }
+
+            foreach ($x in $Tag.Keys)
+            {
+                 if ($x.length -gt 512)
```
Comment (prsadhu-ms-idc) [inline src/Migrate/custom/New-AzMigrateServerReplication.ps1:541]: Check for null/empty?

File: src/Migrate/custom/New-AzMigrateServerReplication.ps1 [hunk 8]
```diff
@@ -464,6 +514,251 @@ public static int hashForArtifact(String artifact)
...
+                 {
+                     $ExceededLengthKey.add($x)
+                 }
+
+                 if ($x -match "[<>%&\?/.]")
+                 {
+                     $IllegalCharKey.add($x)
```
Comment (prsadhu-ms-idc) [inline src/Migrate/custom/New-AzMigrateServerReplication.ps1:548]: throw error here itself?

File: src/Migrate/custom/New-AzMigrateServerReplication.ps1 [hunk 9]
```diff
@@ -464,6 +514,251 @@ public static int hashForArtifact(String artifact)
...
+            }
+
+            foreach ($x in $Tag.Keys)
+            {
+                 if ($x.length -gt 512)
+                 {
+                     $ExceededLengthKey.add($x)
+                 }
+
+                 if ($x -match "[<>%&\?/.]")
```
Comment (prsadhu-ms-idc) [inline src/Migrate/custom/New-AzMigrateServerReplication.ps1:546]: Make the illegal char array a variable, and use it elsewhere. So if we change in one place, it should be good.

File: src/Migrate/custom/New-AzMigrateServerReplication.ps1 [hunk 10]
```diff
@@ -464,6 +514,251 @@ public static int hashForArtifact(String artifact)
...
+            }
+
+            if ($ExceededLengthKey.Count -gt 0)
+            {
+                throw "InvalidTagName : Tag value too large. Following tag value '$($ExceededLengthKey -join ', ')' exceeded the maximum length. Maximum allowed length for tag value - '512' characters."
+            }
+
+            if ($ExceededLengthValue.Count -gt 0)
+            {
+                throw "InvalidTagValueLength : Tag value too large. Following tag value '$($ExceededLengthValue -join ', ')' exceeded the maximum length. Maximum allowed length for tag value - '256' characters."
```
Comment (prsadhu-ms-idc) [inline src/Migrate/custom/New-AzMigrateServerReplication.ps1:618]: Check with Rahul Gupta/Deepika.

File: src/Migrate/custom/Set-AzMigrateServerReplication.ps1 [hunk 11]
```diff
@@ -79,6 +79,70 @@ function Set-AzMigrateServerReplication {
         # Specifies the Availability Zone to be used for VM creation.
         ${TargetAvailabilityZone},
 
+        [Parameter()]
+        [ValidateSet("NoLicenseType" , "PAYG" , "AHUB")]
+        [ArgumentCompleter( { "NoLicenseType" , "PAYG" , "AHUB" })]
+        [Microsoft.Azure.PowerShell.Cmdlets.Migrate.Category('Path')]
+        [System.String]
+        # Specifies if Azure Hybrid benefit is applicable for the source server to migrated.
```
Comment (prsadhu-ms-idc) [inline src/Migrate/custom/Set-AzMigrateServerReplication.ps1:87]: Same as enable.

File: src/Migrate/custom/Set-AzMigrateServerReplication.ps1 [hunk 12]
```diff
@@ -203,6 +286,349 @@ function Set-AzMigrateServerReplication {
                 $ProviderSpecificDetails.TargetAvailabilityZone = $ReplicationMigrationItem.ProviderSpecificDetail.TargetAvailabilityZone
             }
 
+            if ($HasSqlServerLicenseType)
+            {
+                $validLicenseSpellings = @{
```
Comment (prsadhu-ms-idc) [inline src/Migrate/custom/Set-AzMigrateServerReplication.ps1:291]: same as enable.

File: src/Migrate/custom/Set-AzMigrateServerReplication.ps1 [hunk 13]
```diff
@@ -203,6 +286,349 @@ function Set-AzMigrateServerReplication {
...
+                {
+                    throw "InvalidTagName : Tag value too large. Following tag value '$($ExceededLengthKey -join ', ')' exceeded the maximum length. Maximum allowed length for tag value - '512' characters."
+                }
+
+                if ($ExceededLengthValue.Count -gt 0)
+                {
+                    throw "InvalidTagValueLength : Tag value too large. Following tag value '$($ExceededLengthValue -join ', ')' exceeded the maximum length. Maximum allowed length for tag value - '256' characters."
+                }
+
+                if ($UpdateTagsOperation -eq "Merge")
```
Comment (prsadhu-ms-idc) [inline src/Migrate/custom/Set-AzMigrateServerReplication.ps1:343]: check this implementation with Deepika and here: https://docs.microsoft.com/en-us/azure/azure-resource-manager/management/tag-resources

Comment (prsadhu-ms-idc) [reply src/Migrate/custom/Set-AzMigrateServerReplication.ps1:343]: https://stackoverflow.com/questions/25123877/is-there-a-way-to-make-certain-functions-private-in-a-powershell-script/25129094

Comment (prsadhu-ms-idc) [reply src/Migrate/custom/Set-AzMigrateServerReplication.ps1:343]: Explore if we can reduce lines of code by having a dictionary (VmTag: $VmTag, ...) and iterate through it, or maybe private functions that don't get exposed.

File: src/Migrate/custom/Set-AzMigrateServerReplication.ps1 [hunk 14]
```diff
@@ -203,6 +286,349 @@ function Set-AzMigrateServerReplication {
...
+                    if ($ProviderSpecificDetails.TargetDiskTag.Count -gt 50)
+                    {
+                        throw "InvalidTags : Too many tags specified. Requested tag count - '$($ProviderSpecificDetails.TargetDiskTag.Count)'. Maximum number of tags allowed - '50'."
+                    }
+                }
+                else{
+                    $ProviderSpecificDetails.TargetDiskTag = $ReplicationMigrationItem.ProviderSpecificDetail.TargetDiskTag
+                }
+            }
+
```
Comment (prsadhu-ms-idc) [inline src/Migrate/custom/Set-AzMigrateServerReplication.ps1:631]: test case?

**Review by prsadhu-ms-idc — CHANGES_REQUESTED** [review]

File: src/Migrate/custom/New-AzMigrateServerReplication.ps1 (hunk 8 above)
Comment (singhabh27) [reply src/Migrate/custom/New-AzMigrateServerReplication.ps1:548]: It was supposed to print all keys which are failing in this validation. So can't throw from first occurence.

File: src/Migrate/custom/New-AzMigrateServerReplication.ps1 (hunk 9 above)
Comment (singhabh27) [reply src/Migrate/custom/New-AzMigrateServerReplication.ps1:546]: But how can we distinguish the type of error.

File: src/Migrate/custom/New-AzMigrateServerReplication.ps1 (hunk 10 above)
Comment (singhabh27) [reply src/Migrate/custom/New-AzMigrateServerReplication.ps1:618]: Error message is shared by Deepika and it is same as what compute throws.

File: src/Migrate/custom/New-AzMigrateServerReplication.ps1 (hunk 4 above)
Comment (lshai06) [inline src/Migrate/custom/New-AzMigrateServerReplication.ps1:62]: ...for the source server to be migrated.

File: src/Migrate/custom/New-AzMigrateServerReplication.ps1 [hunk 15]
```diff
@@ -119,6 +127,36 @@ function New-AzMigrateServerReplication {
...
+        [Microsoft.Azure.PowerShell.Cmdlets.Migrate.Category('Path')]
+        [Microsoft.Azure.PowerShell.Cmdlets.Migrate.Models.Api20210210.IVMwareCbtEnableMigrationInputSeedDiskTags]
+        # Specifies the tag to be used for Seed Disk creation.
+        ${SeedDiskTag},
+
+        [Parameter()]
+        [Microsoft.Azure.PowerShell.Cmdlets.Migrate.Category('Path')]
+        [System.Collections.Hashtable]
+        # Specifies the tag to be used for Resource creation.
+        ${Tag},
```
Comment (lshai06) [inline src/Migrate/custom/New-AzMigrateServerReplication.ps1:158]: What tag is this resource for? VM, NIC, seed disk, target disk, that's all right?
        /// <summary>
        /// Gets or sets the target VM tags.
        /// </summary>
        [AsrMaxLength(MaxTagCountForVM)]
        [JsonProperty(PropertyName = "targetVmTags")]
        public Dictionary<string, string> TargetVmTags { get; set; }

        /// <summary>
        /// Gets or sets the tags for the seed disks.
        /// </summary>
        [AsrMaxLength(MaxTagCountForDisk - 1)]
        [JsonProperty(PropertyName = "seedDiskTags")]
        public Dictionary<string, string> SeedDiskTags { get; set; }

        /// <summary>
        /// Gets or sets the tags for the target disks.
        /// </summary>
        [AsrMaxLength(MaxTagCountForDisk - 1)]
        [JsonProperty(PropertyName = "targetDiskTags")]
        public Dictionary<string, string> TargetDiskTags { get; set; }

        /// <summary>
        /// Gets or sets the tags for the target NICs.
        /// </summary>
        [AsrMaxLength(MaxTagCountForNic)]
        [JsonProperty(PropertyName = "targetNicTags")]
        public Dictionary<string, string> TargetNicTags { get; set; }

File: src/Migrate/custom/New-AzMigrateServerReplication.ps1 (hunk 3 above)
Comment (lshai06) [reply src/Migrate/custom/New-AzMigrateServerReplication.ps1:139]: Shouldn't this be NICTags? (Plural), as we take a dictionary of tags? For all the tag inputs, same comment.

File: src/Migrate/custom/New-AzMigrateServerReplication.ps1 [hunk 16]
```diff
@@ -119,6 +127,36 @@ function New-AzMigrateServerReplication {
...
+        [Microsoft.Azure.PowerShell.Cmdlets.Migrate.Category('Path')]
+        [Microsoft.Azure.PowerShell.Cmdlets.Migrate.Models.Api20210210.IVMwareCbtEnableMigrationInputTargetNicTags]
+        # Specifies the tag to be used for Nic creation.
+        ${NicTag},
+
+        [Parameter()]
+        [Microsoft.Azure.PowerShell.Cmdlets.Migrate.Category('Path')]
+        [Microsoft.Azure.PowerShell.Cmdlets.Migrate.Models.Api20210210.IVMwareCbtEnableMigrationInputTargetDiskTags]
+        # Specifies the tag to be used for Disk creation.
+        ${DiskTag},
```
Comment (lshai06) [inline src/Migrate/custom/New-AzMigrateServerReplication.ps1:146]: Keep it ~ to service, SeedDiskTags and TargetDiskTags

File: src/Migrate/custom/New-AzMigrateServerReplication.ps1 (hunk 3 above)
Comment (singhabh27) [reply src/Migrate/custom/New-AzMigrateServerReplication.ps1:139]: Kept same as suggested in Resource Tagging Spec_vDraft.docx (sharepoint.com)

File: src/Migrate/custom/New-AzMigrateServerReplication.ps1 [hunk 17]
```diff
@@ -464,6 +514,251 @@ public static int hashForArtifact(String artifact)
...
+            $IllegalCharKey = New-Object Collections.Generic.List[String]
+            $ExceededLengthKey = New-Object Collections.Generic.List[String]
+            $ExceededLengthValue = New-Object Collections.Generic.List[String]
+
+            if ($Tag.Count -gt 50)
+            {
+                throw "InvalidTags : Too many tags specified. Requested tag count - '$($VMTag.Count)'. Maximum number of tags allowed - '50'."
+            }
+
+            foreach ($x in $Tag.Keys)
```
Comment (lshai06) [inline src/Migrate/custom/New-AzMigrateServerReplication.ps1:539]: Don't use x as a variable name.

File: src/Migrate/custom/New-AzMigrateServerReplication.ps1 [hunk 18]
```diff
@@ -464,6 +514,251 @@ public static int hashForArtifact(String artifact)
...
+
+                 if ($($Tag.Item($x)).length -gt 256)
+                 {
+                     $ExceededLengthValue.add($($Tag.Item($x)))
+                 }
+            }
+
+            if ($IllegalCharKey.Count -gt 0)
+            {
+                throw "InvalidTagNameCharacters : The tag names '$($IllegalCharKey -join ', ')' have reserved characters '<,>,%,&,\,?,/' or control characters. These characters are only allowed for tags that start with the prefix 'hidden, link'."
```
Comment (lshai06) [inline src/Migrate/custom/New-AzMigrateServerReplication.ps1:559]: characters are only allowed for tags that start with the prefix 'hidden, link'."
What is hidden, link?

File: src/Migrate/custom/New-AzMigrateServerReplication.ps1 (hunk 15 above)
Comment (singhabh27) [reply src/Migrate/custom/New-AzMigrateServerReplication.ps1:158]: Yes, this tag sets all resources(VM, NIC, Disk) tag values.

Comment (singhabh27) [reply src/Migrate/custom/New-AzMigrateServerReplication.ps1:158]: Yes

File: src/Migrate/custom/New-AzMigrateServerReplication.ps1 (hunk 16 above)
Comment (singhabh27) [reply src/Migrate/custom/New-AzMigrateServerReplication.ps1:146]: Kept same as suggested in PM doc.

File: src/Migrate/custom/Set-AzMigrateServerReplication.ps1 (hunk 13 above)
Comment (singhabh27) [reply src/Migrate/custom/Set-AzMigrateServerReplication.ps1]: Done.

File: src/Migrate/custom/New-AzMigrateServerReplication.ps1 (hunk 18 above)
Comment (singhabh27) [reply src/Migrate/custom/New-AzMigrateServerReplication.ps1]: Removed.

File: src/Migrate/custom/Set-AzMigrateServerReplication.ps1 (hunk 14 above)
Comment (singhabh27) [reply src/Migrate/custom/Set-AzMigrateServerReplication.ps1:485]: Will raise a seperate PR for this.

**Pull request merged**